### PR TITLE
Write catch2 tests

### PR DESCRIPTION
**Pull request by eivindjahren**
Addresses : https://github.com/equinor/ecl/issues/768

### CODE REVIEW

File: lib/tests/test_ecl_grid.cpp [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+  }
+}
+
+class Tmpdir {
+    public:
+        fs::path dirname;
+        Tmpdir() : dirname("") {
+
+            auto parentdir = fs::temp_directory_path() / "ecl_test";
+            dirname = parentdir / to_string(rand());
```
Comment (dotfloat) [inline lib/tests/test_ecl_grid.cpp:255]: do-while

File: lib/tests/test_ecl_grid.cpp [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+            while(fs::exists(dirname)){
+                dirname = parentdir / to_string(rand());
+            }
+            fs::create_directories(dirname);
+
+        }
+        ~Tmpdir() {
+            fs::remove_all(dirname);
```
Comment (dotfloat) [inline lib/tests/test_ecl_grid.cpp:263]: Lets not remove a potentially useful debugging tool. (Maybe check an environment variable, eg. `CLEAN_TMPDIR`)

File: lib/tests/test_ecl_grid.cpp [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    ecl_grid_free(grid);
+  }
+}
+
+class Tmpdir {
+    public:
+        fs::path dirname;
+        Tmpdir() : dirname("") {
+
+            auto parentdir = fs::temp_directory_path() / "ecl_test";
```
Comment (dotfloat) [inline lib/tests/test_ecl_grid.cpp:254]: Output directory to `stderr` to notify the user about the new created tmpdir.

File: lib/tests/test_ecl_grid.cpp (hunk 1 above)
Comment (eivindjahren) [reply lib/tests/test_ecl_grid.cpp:258]: Done

File: lib/tests/test_ecl_grid.cpp (hunk 2 above)
Comment (eivindjahren) [reply lib/tests/test_ecl_grid.cpp]: Now defaults to not remove, commandline option to clean.

File: lib/tests/test_ecl_grid.cpp (hunk 3 above)
Comment (eivindjahren) [reply lib/tests/test_ecl_grid.cpp]: Prints directory to stderr.

File: lib/tests/test_ecl_grid.cpp [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+            REQUIRE(ecl_grid_get_cell_twist3(grid, 0, 0, 0) == 1);
+        }
+
+        ecl_grid_free(grid);
+    }
+}
+
+TEST_CASE("Test double twistedness", "[unittest]") {
+    GIVEN("A grid with the first cell doubly twisted") {
+        auto z_vector = vector<tuple<int, int, int, int, double>>(4);
```
Comment (dotfloat) [inline lib/tests/test_ecl_grid.cpp:229]:
```c++
auto z_vector = vector<tuple<int, int, int, int, double>> {
    {0, 0, 0, 0, 0.0},
    {0, 0, 0, 4, -0.25},
    {0, 0, 0, 3, 0.0},
    {0, 0, 0, 7, -0.25}
};

File: lib/tests/test_ecl_grid.cpp [hunk 5]
```diff
@@ -0,0 +1,321 @@
+#include <catch2/catch.hpp>
+#include <ert/ecl/ecl_grid.hpp>
+#include <ert/ecl/ecl_kw.hpp>
+#include <ert/ecl/ecl_kw_magic.hpp>
+#include <vector>
+
+#include "tmpdir.hpp"
+
+using namespace std;
```
Comment (dotfloat) [inline lib/tests/test_ecl_grid.cpp:9]: Don't `using namespace std;`. Prefer either writing out `std::` or importing each symbol separately using `using std::vector;`

File: lib/tests/test_ecl_grid.cpp [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+ * Generates a grid, using the ecl_grid_alloc_GRDECL_kw constructor, with the
+ * x,y,z coordinates of the 0th corner of the ijkth cel is i,j,k. In other
+ * words this is a grid where the corners are on the whole numbers.
+ *
+ * Takes a vector of i,j,k,c,z tuples which changes the corner c of cell at
+ * i,j,k z value.
+ */
+ecl_grid_type *generate_coordkw_grid(
+    int num_x, int num_y, int num_z,
+    vector<tuple<int, int, int, int, double>> z_vector) {
```
Comment (dotfloat) [inline lib/tests/test_ecl_grid.cpp:144]: You're passing `z_vector` by copy, do it rather by `const &`.

File: lib/tests/test_ecl_grid.cpp [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+                    ecl_kw_iset_float(zcorn_kw, zi2, z2);
+                }
+            }
+        }
+    }
+
+    for (auto &t : z_vector) {
+        int i, j, k, c;
+        double z;
+        tie(i, j, k, c, z) = move(t);
```
Comment (dotfloat) [inline lib/tests/test_ecl_grid.cpp:180]: You're moving objects out of `z_vector`, thus invalidating them. If you ever use any elements of `z_vector` again then it's UB. The correct behaviour here is probably to not `std::move`.

File: lib/tests/test_ecl_grid.cpp [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+                    double z2 = k + 1;
+
+                    ecl_kw_iset_float(zcorn_kw, zi1, z1);
+                    ecl_kw_iset_float(zcorn_kw, zi2, z2);
+                }
+            }
+        }
+    }
+
+    for (auto &t : z_vector) {
```
Comment (dotfloat) [inline lib/tests/test_ecl_grid.cpp:177]: C++17 structured bindings: `for (const auto& [i, j, k, c, z] : z_vector) {`

File: lib/tests/test_ecl_grid.cpp [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+    ecl_kw_free(coord_kw);
+    ecl_kw_free(zcorn_kw);
+
+    return grid;
+}
+
+TEST_CASE("Test no twistedness", "[unittest]") {
+    GIVEN("A grid with untwisted cells") {
+        ecl_grid_type *grid = generate_coordkw_grid(
+            10, 7, 8, vector<tuple<int, int, int, int, double>>(0));
```
Comment (dotfloat) [inline lib/tests/test_ecl_grid.cpp:196]: C++11: replace `vector<tuple<int, int, int, int, double>>(0)` with a single `{}`.

File: lib/tests/test_ecl_grid.cpp [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+            }
+        }
+
+        ecl_grid_free(grid);
+    }
+}
+
+TEST_CASE("Test simple twistedness", "[unittest]") {
+    GIVEN("A grid with the first cell twisted") {
+        auto z_vector = vector<tuple<int, int, int, int, double>>(2);
```
Comment (dotfloat) [inline lib/tests/test_ecl_grid.cpp:214]:
```c++
auto z_vector = vector<tuple<int, int, int, int, double>> {
    {0, 0, 0, 0, 0.0},
    {0, 0, 0, 4, -0.25}
}

**Review by dotfloat — APPROVED** [review]

File: lib/tests/test_ecl_grid.cpp (hunk 4 above)
Comment (eivindjahren) [reply lib/tests/test_ecl_grid.cpp]: Fixed

File: lib/tests/test_ecl_grid.cpp (hunk 5 above)
Comment (eivindjahren) [reply lib/tests/test_ecl_grid.cpp]: Fixed by not doing `using namespace std;`

File: lib/tests/test_ecl_grid.cpp (hunk 6 above)
Comment (eivindjahren) [reply lib/tests/test_ecl_grid.cpp]: Fixed

File: lib/tests/test_ecl_grid.cpp (hunk 8 above)
Comment (eivindjahren) [reply lib/tests/test_ecl_grid.cpp]: Done

File: lib/tests/test_ecl_grid.cpp (hunk 9 above)
Comment (eivindjahren) [reply lib/tests/test_ecl_grid.cpp]: Done

File: lib/tests/test_ecl_grid.cpp (hunk 10 above)
Comment (eivindjahren) [reply lib/tests/test_ecl_grid.cpp]: Done

**Review by dotfloat — APPROVED** [review]
Really nice!!! :tada: :muscle:

**Pull request merged**